Build error and OMP crash on MinGW (64-bit) on Windows 10

See [openfhe discussion: Open FHE Building Error in Windows 10](https://openfhe.discourse.group/t/open-fhe-building-error-in-windows-10/117) for more details.

I was able to recreate the error. Here is an example.

```
[ 46%] Building CXX object src/binfhe/CMakeFiles/boolean-serial-json.dir/examples/boolean-serial-json.cpp.obj
In file included from C:/git/openfhe-development/third-party/google-test/googletest/include/gtest/gtest.h:376,
                 from C:/git/openfhe-development/src/binfhe/unittest/UnitTestFunc.cpp:37:
C:/git/openfhe-development/src/binfhe/unittest/UnitTestFunc.cpp: In member function 'virtual void UnitTestFHEWGINX_EvalArbFunc_Test::TestBody()':
C:/git/openfhe-development/src/binfhe/unittest/UnitTestFunc.cpp:69:19: error: 'uint' was not declared in this scope; did you mean 'usint'?
   69 |         EXPECT_EQ(uint(fp(i, p).ConvertToInt()), result) << failed;
      |                   ^~~~
C:/git/openfhe-development/third-party/google-test/googletest/include/gtest/gtest_pred_impl.h:77:52: note: in definition of macro 'GTEST_ASSERT_'
   77 |   if (const ::testing::AssertionResult gtest_ar = (expression)) \
      |                                                    ^~~~~~~~~~
C:/git/openfhe-development/third-party/google-test/googletest/include/gtest/gtest_pred_impl.h:164:3: note: in expansion of macro 'GTEST_PRED_FORMAT2_'
  164 |   GTEST_PRED_FORMAT2_(pred_format, v1, v2, GTEST_NONFATAL_FAILURE_)
      |   ^~~~~~~~~~~~~~~~~~~
C:/git/openfhe-development/third-party/google-test/googletest/include/gtest/gtest.h:2028:3: note: in expansion of macro 'EXPECT_PRED_FORMAT2'
 2028 |   EXPECT_PRED_FORMAT2(::testing::internal::EqHelper::Compare, val1, val2)
      |   ^~~~~~~~~~~~~~~~~~~
C:/git/openfhe-development/src/binfhe/unittest/UnitTestFunc.cpp:69:9: note: in expansion of macro 'EXPECT_EQ'
   69 |         EXPECT_EQ(uint(fp(i, p).ConvertToInt()), result) << failed;
      |         ^~~~~~~~~
C:/git/openfhe-development/src/binfhe/unittest/UnitTestFunc.cpp: In member function 'virtual void UnitTestFHEWGINX_EvalFloorFunc_Test::TestBody()':
C:/git/openfhe-development/src/binfhe/unittest/UnitTestFunc.cpp:95:19: error: 'uint' was not declared in this scope; did you mean 'usint'?
   95 |         EXPECT_EQ(uint(i / 2), result) << failed;
      |                   ^~~~
C:/git/openfhe-development/third-party/google-test/googletest/include/gtest/gtest_pred_impl.h:77:52: note: in definition of macro 'GTEST_ASSERT_'
   77 |   if (const ::testing::AssertionResult gtest_ar = (expression)) \
      |                                                    ^~~~~~~~~~
C:/git/openfhe-development/third-party/google-test/googletest/include/gtest/gtest_pred_impl.h:164:3: note: in expansion of macro 'GTEST_PRED_FORMAT2_'
  164 |   GTEST_PRED_FORMAT2_(pred_format, v1, v2, GTEST_NONFATAL_FAILURE_)
      |   ^~~~~~~~~~~~~~~~~~~
C:/git/openfhe-development/third-party/google-test/googletest/include/gtest/gtest.h:2028:3: note: in expansion of macro 'EXPECT_PRED_FORMAT2'
 2028 |   EXPECT_PRED_FORMAT2(::testing::internal::EqHelper::Compare, val1, val2)
      |   ^~~~~~~~~~~~~~~~~~~
C:/git/openfhe-development/src/binfhe/unittest/UnitTestFunc.cpp:95:9: note: in expansion of macro 'EXPECT_EQ'
   95 |         EXPECT_EQ(uint(i / 2), result) << failed;
      |         ^~~~~~~~~
C:/git/openfhe-development/src/binfhe/unittest/UnitTestFunc.cpp: In member function 'virtual void UnitTestFHEWGINX_EvalSignFuncTime_Test::TestBody()':
C:/git/openfhe-development/src/binfhe/unittest/UnitTestFunc.cpp:118:19: error: 'uint' was not declared in this scope; did you mean 'usint'?
  118 |         EXPECT_EQ(uint(i >= 3), result) << failed;
```

I was able to replicate it too and am working to fix it.

On Wed, Sep 7, 2022 at 11:33 AM yspolyakov ***@***.***> wrote:

> I was able to recreate the error. Here is an example.
>
> [ 46%] Building CXX object src/binfhe/CMakeFiles/boolean-serial-json.dir/examples/boolean-serial-json.cpp.obj
> In file included from C:/git/openfhe-development/third-party/google-test/googletest/include/gtest/gtest.h:376,
>                  from C:/git/openfhe-development/src/binfhe/unittest/UnitTestFunc.cpp:37:
> C:/git/openfhe-development/src/binfhe/unittest/UnitTestFunc.cpp: In member function 'virtual void UnitTestFHEWGINX_EvalArbFunc_Test::TestBody()':
> C:/git/openfhe-development/src/binfhe/unittest/UnitTestFunc.cpp:69:19: error: 'uint' was not declared in this scope; did you mean 'usint'?
>    69 |         EXPECT_EQ(uint(fp(i, p).ConvertToInt()), result) << failed;
>       |                   ^~~~
> C:/git/openfhe-development/third-party/google-test/googletest/include/gtest/gtest_pred_impl.h:77:52: note: in definition of macro 'GTEST_ASSERT_'
>    77 |   if (const ::testing::AssertionResult gtest_ar = (expression)) \
>       |                                                    ^~~~~~~~~~
> C:/git/openfhe-development/third-party/google-test/googletest/include/gtest/gtest_pred_impl.h:164:3: note: in expansion of macro 'GTEST_PRED_FORMAT2_'
>   164 |   GTEST_PRED_FORMAT2_(pred_format, v1, v2, GTEST_NONFATAL_FAILURE_)
>       |   ^~~~~~~~~~~~~~~~~~~
> C:/git/openfhe-development/third-party/google-test/googletest/include/gtest/gtest.h:2028:3: note: in expansion of macro 'EXPECT_PRED_FORMAT2'
>  2028 |   EXPECT_PRED_FORMAT2(::testing::internal::EqHelper::Compare, val1, val2)
>       |   ^~~~~~~~~~~~~~~~~~~
> C:/git/openfhe-development/src/binfhe/unittest/UnitTestFunc.cpp:69:9: note: in expansion of macro 'EXPECT_EQ'
>    69 |         EXPECT_EQ(uint(fp(i, p).ConvertToInt()), result) << failed;
>       |         ^~~~~~~~~
> C:/git/openfhe-development/src/binfhe/unittest/UnitTestFunc.cpp: In member function 'virtual void UnitTestFHEWGINX_EvalFloorFunc_Test::TestBody()':
> C:/git/openfhe-development/src/binfhe/unittest/UnitTestFunc.cpp:95:19: error: 'uint' was not declared in this scope; did you mean 'usint'?
>    95 |         EXPECT_EQ(uint(i / 2), result) << failed;
>       |                   ^~~~
> C:/git/openfhe-development/third-party/google-test/googletest/include/gtest/gtest_pred_impl.h:77:52: note: in definition of macro 'GTEST_ASSERT_'
>    77 |   if (const ::testing::AssertionResult gtest_ar = (expression)) \
>       |                                                    ^~~~~~~~~~
> C:/git/openfhe-development/third-party/google-test/googletest/include/gtest/gtest_pred_impl.h:164:3: note: in expansion of macro 'GTEST_PRED_FORMAT2_'
>   164 |   GTEST_PRED_FORMAT2_(pred_format, v1, v2, GTEST_NONFATAL_FAILURE_)
>       |   ^~~~~~~~~~~~~~~~~~~
> C:/git/openfhe-development/third-party/google-test/googletest/include/gtest/gtest.h:2028:3: note: in expansion of macro 'EXPECT_PRED_FORMAT2'
>  2028 |   EXPECT_PRED_FORMAT2(::testing::internal::EqHelper::Compare, val1, val2)
>       |   ^~~~~~~~~~~~~~~~~~~
> C:/git/openfhe-development/src/binfhe/unittest/UnitTestFunc.cpp:95:9: note: in expansion of macro 'EXPECT_EQ'
>    95 |         EXPECT_EQ(uint(i / 2), result) << failed;
>       |         ^~~~~~~~~
> C:/git/openfhe-development/src/binfhe/unittest/UnitTestFunc.cpp: In member function 'virtual void UnitTestFHEWGINX_EvalSignFuncTime_Test::TestBody()':
> C:/git/openfhe-development/src/binfhe/unittest/UnitTestFunc.cpp:118:19: error: 'uint' was not declared in this scope; did you mean 'usint'?
>   118 |         EXPECT_EQ(uint(i >= 3), result) << failed;
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openfheorg/openfhe-development/issues/176#issuecomment-1239548556>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AG46TLR363XDFKBOXQARMIDV5CYWFANCNFSM6AAAAAAQGBSXEE>
> .
> You are receiving this because you were assigned.Message ID:
> ***@***.***>
>

--- a/src/binfhe/unittest/UnitTestFunc.cpp
+++ b/src/binfhe/unittest/UnitTestFunc.cpp
@@ -66,7 +66,7 @@ TEST(UnitTestFHEWGINX, EvalArbFunc) {
         cc.Decrypt(sk, ct_cube, &result, p);
 
         std::string failed = "Arbitrary Function Evaluation failed";
-        EXPECT_EQ(uint(fp(i, p).ConvertToInt()), result) << failed;
+        EXPECT_EQ(usint(fp(i, p).ConvertToInt()), result) << failed;
     }
 }
 
@@ -92,7 +92,7 @@ TEST(UnitTestFHEWGINX, EvalFloorFunc) {
         cc.Decrypt(sk, ctRounded, &result, p / 2);
 
         std::string failed = "Floor Function Evalution failed";
-        EXPECT_EQ(uint(i / 2), result) << failed;
+        EXPECT_EQ(usint(i / 2), result) << failed;
     }
 }
 
@@ -115,7 +115,7 @@ TEST(UnitTestFHEWGINX, EvalSignFuncTime) {
         ct1      = cc.EvalSign(ct1, Q);
         LWEPlaintext result;
         cc.Decrypt(sk, ct1, &result, 2, q);
-        EXPECT_EQ(uint(i >= 3), result) << failed;
+        EXPECT_EQ(usint(i >= 3), result) << failed;
     }
 }
 
@@ -138,7 +138,7 @@ TEST(UnitTestFHEWGINX, EvalSignFuncSpace) {
         ct1      = cc.EvalSign(ct1, Q);
         LWEPlaintext result;
         cc.Decrypt(sk, ct1, &result, 2, q);
-        EXPECT_EQ(uint(i >= 3), result) << failed;
+        EXPECT_EQ(usint(i >= 3), result) << failed;
     }
 }
 
@@ -166,10 +166,10 @@ TEST(UnitTestFHEWGINX, EvalDigitDecompTime) {
         LWEPlaintext result;
         cc.Decrypt(sk, ct1, &result, p_basic, basic);
         if (i < st + 3)
-            EXPECT_EQ(uint(15), result) << failed;
+            EXPECT_EQ(usint(15), result) << failed;
         else
-            EXPECT_EQ(uint(0), result) << failed;
-        EXPECT_EQ(uint(ceil(log(factor) / log(p_basic)) + 1), decomp.size()) << failed;
+            EXPECT_EQ(usint(0), result) << failed;
+        EXPECT_EQ(usint(ceil(log(factor) / log(p_basic)) + 1), decomp.size()) << failed;
     }
 }
 
@@ -197,9 +197,9 @@ TEST(UnitTestFHEWGINX, EvalDigitDecompSpace) {
         LWEPlaintext result;
         cc.Decrypt(sk, ct1, &result, p_basic, basic);
         if (i < st + 3)
-            EXPECT_EQ(uint(15), result) << failed;
+            EXPECT_EQ(usint(15), result) << failed;
         else
-            EXPECT_EQ(uint(0), result) << failed;
-        EXPECT_EQ(uint(ceil(log(factor) / log(p_basic)) + 1), decomp.size()) << failed;
+            EXPECT_EQ(usint(0), result) << failed;
+        EXPECT_EQ(usint(ceil(log(factor) / log(p_basic)) + 1), decomp.size()) << failed;
     }
 }

--- a/src/core/include/math/distributiongenerator.h
+++ b/src/core/include/math/distributiongenerator.h
@@ -128,8 +128,8 @@ public:
                 // location as the counter for BLAKE2 This will increase the entropy of
                 // the BLAKE2 sample
                 void* mem = malloc(1);
-                free(mem);
                 uint32_t counter = reinterpret_cast<long long>(mem);  // NOLINT
+                free(mem);
 
                 PRNG gen(initKey, counter);
 

--- a/src/pke/lib/scheme/ckksrns/ckksrns-fhe.cpp
+++ b/src/pke/lib/scheme/ckksrns/ckksrns-fhe.cpp
@@ -844,7 +844,11 @@ std::vector<ConstPlaintext> FHECKKSRNS::EvalLinearTransformPrecompute(
     //  auto elementParamsPtr2 = std::dynamic_pointer_cast<typename DCRTPoly::Params>(elementParamsPtr);
 
     std::vector<ConstPlaintext> result(slots);
-#pragma omp parallel for
+// parallelizing the loop (below) with OMP causes a segfault on MinGW
+// see https://github.com/openfheorg/openfhe-development/issues/176
+#if !defined(__MINGW32__) && !defined(__MINGW64__)
+    #pragma omp parallel for
+#endif
     for (int j = 0; j < gStep; j++) {
         int offset = -bStep * j;
         for (int i = 0; i < bStep; i++) {

--- a/src/pke/lib/scheme/ckksrns/ckksrns-fhe.cpp
+++ b/src/pke/lib/scheme/ckksrns/ckksrns-fhe.cpp
@@ -1074,7 +1074,9 @@ std::vector<std::vector<ConstPlaintext>> FHECKKSRNS::EvalCoeffsToSlotsPrecompute
 
         for (int32_t s = levelBudget - 1; s > stop; s--) {
             for (int32_t i = 0; i < b; i++) {
-#pragma omp parallel for
+#if !defined(__MINGW32__) && !defined(__MINGW64__)
+    #pragma omp parallel for
+#endif
                 for (int32_t j = 0; j < g; j++) {
                     if (g * i + j != int32_t(numRotations)) {
                         uint32_t rot =
@@ -1123,7 +1125,9 @@ std::vector<std::vector<ConstPlaintext>> FHECKKSRNS::EvalCoeffsToSlotsPrecompute
 
         for (int32_t s = levelBudget - 1; s > stop; s--) {
             for (int32_t i = 0; i < b; i++) {
-#pragma omp parallel for
+#if !defined(__MINGW32__) && !defined(__MINGW64__)
+    #pragma omp parallel for
+#endif
                 for (int32_t j = 0; j < g; j++) {
                     if (g * i + j != int32_t(numRotations)) {
                         uint32_t rot =

